keps/sig-network/3453-minimize-iptables-restore/README.md

```diff
@@ -0,0 +1,468 @@
+small additional performance improvement relative to the main
+improvement of not rewriting _every_ service and endpoint rule on
+_every_ resync.
+
+In the event that a partial `iptables-restore` call fails, kube-proxy
+will queue another resync, and fall back to doing a full
+`iptables-restore` that time. (Note that there are no expected cases
+where a partial restore would fail but a full restore would succeed;
+this would always indicate a bug in the code.) Likewise, it always
+does a full resync when there are topology-related changes to Node
+labels, and it always does a full resync at least once every
+`iptablesSyncPeriod`.
+
+[the proposed PR]: https://github.com/kubernetes/kubernetes/pull/110268
+
+### Risks and Mitigations
+
+#### Catastrophic Failure
+
+In the event of bugs in the code, kube-proxy could fail
+catastrophically, entirely breaking the cluster. This would be "bad".
+
+#### Subtle Optimization Failure
+
+The code could also fail more subtly. In the worst case, every attempt
+to do a partial restore might fail, causing the code to fall back
+to the slower full `iptables-restore` on the next sync, with the end
+result that kube-proxy becomes _less_ efficient than the old code that
+would have just done a full restore the first time.
+
+Less obviously, the new code might reliably cause a fail-and-retry
+only in certain circumstances (eg, when a service goes from 1 endpoint
+to 0, or when a particular Service property changes).
+
+Since, as mentioned above, there are no _expected_ cases where a
+partial restore will fail, we could notice that this failure mode was
+happening by having a metric for how often a partial restore fails,
+and if we find that it is happening (in e2e tests or real clusters),
+we can then debug or revert the bad code.
+
+#### Subtle Synchronization Delays
+
+The new code needs to analyze Service/EndpointSlice changes to
+determine when the service-specific chains for each service need to be
+included in the `iptables-restore` input. If this analysis is
+incorrect, it might fail to update the iptables rules for a service in
+response to certain kinds of changes (eg, when a service goes from 1
+endpoint to 0, or when a particular Service property changes), causing
+the iptables rules to be stale until they eventually get fixed in a
+future full resync.
+
+In order to detect this, we would have to sanity-check every partial
+restore, to ensure that, if applied to the current kernel state, it
+would generate the same end result as the full restore we would
+otherwise have applied. Although we could do this with the
+`FakeIPTables` code in `pkg/util/iptables/testing`, that code is only
+intended for the sorts of small-ish test cases in the unit tests, and
+not at all optimized for handling the sort of very large rule sets
+that would exist in larger clusters (ie, the rule sets that would
+benefit the most from the partial restore feature).
+
+One possible approach for dealing with that would be to run the
```

Comment:
I actually really like this idea.  Could we generate both the dump and the patch, apply the patch, then read it back and ensure that it matches the full dump?  If it fails, set a metric and fall back on full-dump for the future.  It would be heavyweight but we only need to do it when the gate is on.

Reply:
@danwinship Can we pin down the plan and convert words from "possible" to "will" ?

Reply:
Having tried to figure out how I would implement this, I'm now pretty convinced this is a bad idea. The "checking" code would be much much much more complicated than the code that it is checking, meaning that if it ever triggered, it would most likely indicate a bug in the checking code, not an actual bug in the syncing code. (And recent events (#112477) show that iptables upstream does not consider `iptables-save` output to be "stable-API-like", so there's yet another source of flakes if we're trying to parse that.)

Also, _we have no reason to believe that this failure mode will actually occur_. I was just trying to come up with all of the ways that the code could fail, and this is one of them. But it's not an especially plausible one. Much more likely is that if there was a bug, it would involve one of the existing code branches in `syncProxyRules` (eg, we don't sync correctly if only the firewall rules change) and it would be caught very quickly.